add numeric type hierarchy and issubdtype as well as a set_dtype meth…

dastrobu opened the pull request:
…od to nn.Module with predicate

numeric type hierarchy and issubtype is compatible to the [numpy hierarchy](https://github.com/numpy/numpy/blob/220f0ab2c5958f9d49b9a28c69a760657b06d9b9/numpy/_core/numerictypes.py#L42).

Closes #285.

## Proposed changes

Please include a description of the problem or feature this PR is addressing. If there is a corresponding issue, include the issue #.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented:
Thanks for adding this!

I think exposing an `issubtype` is nice. But I would prefer we do it in the C++ core rather than in the bindings. That way we can use it in C++ and elsewhere. Also we have a lot of information there already that I think you could use to substantially simplify your implementation. Check these files:

https://github.com/ml-explore/mlx/blob/main/mlx/dtype.h
https://github.com/ml-explore/mlx/blob/main/mlx/dtype.cpp

awni commented:
@dastrobu are you planning to come back to this? Would love to see this added as a feature, I think it would be really useful!

dastrobu commented:
@awni yes I do. 
I couldn't allocate time due to other work commitments. Optimistically, I aim to address your comments next week and push this pr forward.

awni commented:
Cool, thanks for the update! Looking forward to it.

awni commented:
@dastrobu this PR is pretty old. Do you intend to come back to it or should we close?

dastrobu commented:
> @dastrobu this PR is pretty old. Do you intend to come back to it or should we close?

@awni,I apologize that my previous "ready for review" marking may have gone unnoticed. I've rebased the pull request against the current main branch and addressed your feedback on the initial draft. I'd appreciate it if you could take another look.

awni commented:
Of course, I will gladly look!

awni commented on python/mlx/nn/layers/base.py:
```diff
@@ -559,3 +559,24 @@ def eval(self):
+        """Set the dtype of the models parameters.
+
+        Args:
+            dtype (Dtype): the new dtype
```
```suggestion
            dtype (Dtype): The new dtype
```

awni commented on python/mlx/nn/layers/base.py:
```diff
@@ -559,3 +559,24 @@ def eval(self):
+
+        Args:
+            dtype (Dtype): the new dtype
+            predicate (~typing.Callable): a predicate to select parameters. By default, only parameters of type
+                :attr:`~mlx.core.floating` will be updated to avoid updating any integer parameters to the new dtype.
```
```suggestion
            predicate (~typing.Callable): A predicate to select parameters to cast. By default, only parameters of type
                :attr:`~mlx.core.floating` will be updated to avoid updating integer parameters to the new dtype.
```

awni commented on python/tests/test_ops.py:
```diff
@@ -2011,6 +2011,40 @@ def compare_nested_lists(x, y):
             self.assertTrue(mx.all(mx.equal(mx_res, atleast_arrays[i])))
 
+    def test_issubdtype(self):
+        self.assertTrue(mx.issubdtype(mx.bfloat16, mx.inexact))
```
Why is `bfloat16` an `inexact` type? Should it not be floating point?

awni commented on docs/src/python/data_types.rst:
```diff
@@ -50,3 +50,18 @@ The default floating point type is ``float32`` and the default integer type is
+.. autosummary::
+  :toctree: _autosummary
+
+  complexfloating
+  floating
+  generic
+  inexact
+  integer
+  number
+  signedinteger
+  unsignedinteger
```
The doc here is not that great, it just has the same line next to each:

<img width="813" alt="Screenshot 2024-03-01 at 2 30 37 PM" src="https://github.com/ml-explore/mlx/assets/1542805/8eb2b584-010d-48c2-8f2d-e145c23f5561">


Maybe we could do it as a table with a line below `Categories` explaining that these are the :obj:`DtypeCategory`? (Kind of like how we did it with the dtypes above).

dastrobu replied:
Good point. I'd suggest to simply embed the docstring of `DtypeCategory`, which documents the hierarchy already  to avoid duplicating the information. What do you think: 

![image](https://github.com/ml-explore/mlx/assets/1847260/870ce2e7-3081-4d62-a4b6-ee9bc4704d5e)

awni replied:
This sounds good to me!

awni reviewed (commented):
@dastrobu this is really nicely done! And thanks for the nice updates to the documentation.

One question I have is that in C++ we now have two ways to get the same info: `is_floating_point(x.dtype())` and `issubdtype(x.dtype(), Dtype::Category::floating)`. That is not so bad I suppose, though it would be cleaner to have only one, but I am a little concerned by the fact that they have different semantics (one includes `complex`, the other does not). Wdyt?

dastrobu commented:
> One question I have is that in C++ we now have two ways to get the same info: `is_floating_point(x.dtype())` and `issubdtype(x.dtype(), Dtype::Category::floating)`. That is not so bad I suppose, though it would be cleaner to have only one, but I am a little concerned by the fact that they have different semantics (one includes `complex`, the other does not). Wdyt?

I agree that this is a bit confusing. The implementation of `issubtype` follows the exact hierarchy of numpy, which makes sense for the python API. If I am not mistaken `is_floating_point` is not exposed to python - so there is no issue for the python API. 

The C++ API is a bit confusing now, I agree. One option would be to remove `is_floating_point` (and potentially `is_unsigned`, ...) in favor of `issubdtype`. I need to take a closer look what that would mean when I have time. 
At least we should update the docs for `is_floating_point ` to clarify the relation to `issubdtype`. I'll let you know when I found the time for a deeper analysis.

awni commented:
> The C++ API is a bit confusing now, I agree. One option would be to remove is_floating_point (and potentially is_unsigned, ...) in favor of issubdtype. I need to take a closer look what that would mean when I have time.
At least we should update the docs for is_floating_point  to clarify the relation to issubdtype. I'll let you know when I found the time for a deeper analysis.

Yes exactly.. it is a larger diff though. A comment for now would be good. We don't have to block merging this on that.

dastrobu commented:
> ... A comment for now would be good. We don't have to block merging this on that.

Unfortunately, it is not only `is_floating_point` which behaves differently... I added a comment on all of the `is_...` functions, explaining the relation to `issubdtype`.

PR is rebased onto current main and would be ready to merge from my side.

abdussamettrkr commented:
Hello, any update here? I am just about to open an issue regarding 'issubtype'. I believe it's addressed here.

awni commented:
Sorry I think this is waiting on us to review. It also needs to be rebased post-nanobind.. @dastrobu let me know if you'd like assistance there.

dastrobu commented:
@awni rebasing is done. I just had to rework `py:pos_only()` which is now `""_a`.

awni reviewed (approved):
Thanks for the changes here! 

I made a few adjustments:
- Change the data structure we use for checking the category. Fixes linux build, and faster
- Go ahead and remove/replace `is_floating` and friends to unify method for checking categories
- Updated docs a bit
